File: 143_Minimum Path Sum.py
"""
143. Minimum Path Sum
DP across rows keeping rolling 1D array.
"""
from typing import List

def min_path_sum(grid: List[List[int]]) -> int:
    m, n = len(grid), len(grid[0])
    dp = [float('inf')] * n
    dp[0] = 0
    for i in range(m):
        for j in range(n):
            if j == 0:
                dp[j] = dp[j] + grid[i][j]
            else:
                dp[j] = min(dp[j], dp[j-1]) + grid[i][j]
    return dp[-1]

if __name__ == '__main__':
    print('minPathSum =', min_path_sum([[1,3,1],[1,5,1],[4,2,1]]))

# Time Complexity: O(m*n)
# Space Complexity: O(n)

File: 143_Minimum Path Sum.cpp
/*
143. Minimum Path Sum
Given a grid, find min path sum from top-left to bottom-right (only right and down moves).
Use in-place DP or separate dp array.
*/
#include <bits/stdc++.h>
using namespace std;

int minPathSum(vector<vector<int>>& grid) {
    int m = grid.size();
    int n = grid[0].size();
    vector<int> dp(n, INT_MAX);
    dp[0] = 0;
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            if (j == 0) dp[j] = dp[j] + grid[i][j];
            else dp[j] = min(dp[j], dp[j-1]) + grid[i][j];
        }
    }
    return dp[n-1];
}

int main() {
    vector<vector<int>> g{{1,3,1},{1,5,1},{4,2,1}};
    cout << "minPathSum = " << minPathSum(g) << "\n"; // 7
    return 0;
}

// Time Complexity: O(m*n)
// Space Complexity: O(n)
